File: internal/api/response.go
package api

import (
	"encoding/json"
	"errors"
	"net/http"

	"github.com/Z-M-Huang/openhive/internal/domain"
)

// successResponse wraps a data payload in a success response.
type successResponse struct {
	Data interface{} `json:"data"`
}

// errorResponse wraps an error in a structured response.
type errorResponse struct {
	Error errorBody `json:"error"`
}

type errorBody struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// JSON writes a JSON success response.
func JSON(w http.ResponseWriter, status int, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(successResponse{Data: data})
}

// Error writes a JSON error response.
func Error(w http.ResponseWriter, status int, code string, message string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(errorResponse{
		Error: errorBody{Code: code, Message: message},
	})
}

// MapDomainError maps domain errors to HTTP status codes and writes the response.
func MapDomainError(w http.ResponseWriter, err error) {
	var nfe *domain.NotFoundError
	var ve *domain.ValidationError
	var ce *domain.ConflictError
	var ele *domain.EncryptionLockedError
	var rle *domain.RateLimitedError

	switch {
	case errors.As(err, &nfe):
		Error(w, http.StatusNotFound, nfe.Code(), "the requested resource was not found")
	case errors.As(err, &ve):
		Error(w, http.StatusBadRequest, ve.Code(), ve.Error())
	case errors.As(err, &ce):
		Error(w, http.StatusConflict, ce.Code(), "a resource conflict occurred")
	case errors.As(err, &ele):
		Error(w, http.StatusForbidden, ele.Code(), "encryption is locked")
	case errors.As(err, &rle):
		w.Header().Set("Retry-After", itoa(rle.RetryAfterSeconds))
		Error(w, http.StatusTooManyRequests, rle.Code(), "rate limit exceeded")
	default:
		Error(w, http.StatusInternalServerError, "INTERNAL_ERROR", "an internal error occurred")
	}
}

func itoa(i int) string {
	if i < 0 {
		return "-" + itoa(-i)
	}
	if i < 10 {
		return string(rune('0' + i))
	}
	return itoa(i/10) + string(rune('0'+i%10))
}

File: backend/src/api/response.ts
/**
 * OpenHive Backend - API Response Helpers
 *
 * JSON success/error response helpers and domain error → HTTP status mapping.
 * Wire format:
 *   Success: { "data": <payload> }
 *   Error:   { "error": { "code": "...", "message": "..." } }
 *
 */

import {
  NotFoundError,
  ValidationError,
  ConflictError,
  EncryptionLockedError,
  RateLimitedError,
} from '../domain/errors.js';

// ---------------------------------------------------------------------------
// Fastify shim types
// ---------------------------------------------------------------------------

/** Minimal Fastify reply interface needed by response helpers. */
export interface FastifyReplyShim {
  code(statusCode: number): FastifyReplyShim;
  header(name: string, value: string): FastifyReplyShim;
  send(payload: unknown): void;
}

// ---------------------------------------------------------------------------
// Wire types
// ---------------------------------------------------------------------------

interface SuccessEnvelope<T> {
  data: T;
}

interface ErrorBody {
  code: string;
  message: string;
}

interface ErrorEnvelope {
  error: ErrorBody;
}

// ---------------------------------------------------------------------------
// sendJSON — success response
// ---------------------------------------------------------------------------

/**
 * Writes a JSON success response wrapped in { data: ... }.
 */
export function sendJSON<T>(reply: FastifyReplyShim, status: number, data: T): void {
  const envelope: SuccessEnvelope<T> = { data };
  reply.code(status).send(envelope);
}

// ---------------------------------------------------------------------------
// sendError — error response
// ---------------------------------------------------------------------------

/**
 * Writes a JSON error response wrapped in { error: { code, message } }.
 */
export function sendError(
  reply: FastifyReplyShim,
  status: number,
  code: string,
  message: string,
): void {
  const envelope: ErrorEnvelope = { error: { code, message } };
  reply.code(status).send(envelope);
}

// ---------------------------------------------------------------------------
// mapDomainError — domain error → HTTP status
// ---------------------------------------------------------------------------

/**
 * Maps domain errors to appropriate HTTP status codes and writes the response.
 * Mapping:
 *   NotFoundError         → 404 NOT_FOUND
 *   ValidationError       → 400 VALIDATION_ERROR
 *   ConflictError         → 409 CONFLICT
 *   EncryptionLockedError → 403 ENCRYPTION_LOCKED
 *   RateLimitedError      → 429 RATE_LIMITED (+ Retry-After header)
 *   default               → 500 INTERNAL_ERROR
 */
export function mapDomainError(reply: FastifyReplyShim, err: unknown): void {
  if (err instanceof NotFoundError) {
    sendError(reply, 404, err.code, 'the requested resource was not found');
    return;
  }

  if (err instanceof ValidationError) {
    sendError(reply, 400, err.code, err.message);
    return;
  }

  if (err instanceof ConflictError) {
    sendError(reply, 409, err.code, 'a resource conflict occurred');
    return;
  }

  if (err instanceof EncryptionLockedError) {
    sendError(reply, 403, err.code, 'encryption is locked');
    return;
  }

  if (err instanceof RateLimitedError) {
    reply.header('Retry-After', String(err.retryAfterSeconds));
    sendError(reply, 429, err.code, 'rate limit exceeded');
    return;
  }

  sendError(reply, 500, 'INTERNAL_ERROR', 'an internal error occurred');
}

// ---------------------------------------------------------------------------
// maskSecret — redact sensitive strings for display
// ---------------------------------------------------------------------------

/**
 * Returns a masked version of a secret string: '****' + last 4 characters.
 * If the string is 4 characters or shorter, returns '****'.
 *
 * Example: maskSecret('my-api-key-xyz') → '****-xyz'
 * Example: maskSecret('abcd') → '****'
 */
export function maskSecret(s: string): string {
  if (s.length <= 4) {
    return '****';
  }
  return '****' + s.slice(-4);
}
